### PR TITLE
feat(cli): integrate D1 with "migrate diff" and "db pull"

### PR DESCRIPTION
**Pull request by jkomyno**
This PR:
- closes https://github.com/prisma/team-orm/issues/1062
- closes https://github.com/prisma/team-orm/issues/1063
- deprecates https://github.com/prisma/prisma/pull/23613

In particular, it:
- adds the `--from-local-d1` and `--to-local-d1` boolean flags to `prisma migrate diff`. They locate the local Cloudflare D1 database automatically, if it exists.
- adds the `--local-d1` boolean flag to `prisma db pull`. It locate the local Cloudflare D1 database automatically, if it exists.
  - When `schema.prisma` exists already, it overrides it. If the schema had no `"driverAdapters"` preview feature, the following warning is emitted:
    ```
    Without the driverAdapters preview feature, the schema introspected via the --local-d1 flag will not work with @prisma/client.
    ```
    Screenshot 2024-03-29 at 17 40 14

  - When no `schema.prisma` exists, it generates it in `./schema.prisma` (rather than `./prisma/schema.prisma`, matching the same location used by `--url`), including a `generator client` block with the `"driverAdapters"` preview feature.
  - An introspected datasource looks like:
    ```prisma
    datasource db {
      provider = "sqlite"
      url      = "file:../.wrangler/state/v3/d1/miniflare-D1DatabaseObject/5d11bcce386042472d19a6a4f58e40041ebc5932c972e1449cbf404f3e3c4a7a.sqlite"
    }
    ```
- adds argument validation
- adds tests for `prisma migrate diff`
- adds tests for `prisma db pull`
- I've also manually tested this PR against a Cloudflare Worker

---

## How is a local Cloudflare D1 database located?

Let's see an example when using `migrate diff --from-local-d1`:

1. Prisma looks into `.wrangler/state/v3/d1/miniflare-D1DatabaseObject/<UUID>.sqlite`.
2. If no `.sqlite` file exists, the following error is emitted:
  ```
  No Cloudflare D1 databases found in .wrangler/state/v3/d1/miniflare-D1DatabaseObject/<UUID>.sqlite.
  Did you run `wrangler d1 create <DATABASE_NAME>` and `wrangler dev`?
  ```
3. If multiple `.sqlite` files exist, the following error are emitted:
  ```
  Multiple Cloudflare D1 databases found in .wrangler/state/v3/d1/miniflare-D1DatabaseObject.
  Please manually specify the local D1 database with `--from-url file:`, without using the `--from-local-d1` flag.`,
  ```

This logic is shared between `prisma migrate diff` and `prisma db pull`, and lives in `packages/internals/src/utils/cloudflareD1.ts`.

### CODE REVIEW

File: packages/migrate/src/__tests__/DbPull/__snapshots__/sqlite.test.ts.snap [hunk 1]
```diff
@@ -163,6 +163,36 @@ model AwesomeProfile {
...
 
+exports[`common/sqlite should succeed when --local-d1 and a single local Cloudflare D1 database exists 2`] = `
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["driverAdapters"]
+}
+
+datasource db {
+  provider = "sqlite"
+  url      = "file:.wrangler/state/v3/d1/miniflare-D1DatabaseObject/5d11bcce386042472d19a6a4f58e40041ebc5932c972e1449cbf404f3e3c4a7a.sqlite"
```
Comment (Jolg42) [inline packages/migrate/src/__tests__/DbPull/__snapshots__/sqlite.test.ts.snap:174]: I'm not sure if that snapshot will be reliable with this value.

Comment (jkomyno) [reply packages/migrate/src/__tests__/DbPull/__snapshots__/sqlite.test.ts.snap:174]: The snapshots is reliable, as the `packages/migrate/src/__tests__/fixtures/cloudflare-d1-many-dbs/.wrangler/state/v3/d1/miniflare-D1DatabaseObject/5d11bcce386042472d19a6a4f58e40041ebc5932c972e1449cbf404f3e3c4a7a.sqlite` database has a fixed `5d11bcce386042472d19a6a4f58e40041ebc5932c972e1449cbf404f3e3c4a7a.sqlite` name :)

File: packages/internals/src/utils/cloudflareD1.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// where `<UUID>` is a unique identifier for the database.
+export async function locateLocalCloudflareD1({ arg }: TocateLocalCloudflareD1Args) {
+  const cwd = process.cwd()
+  const d1DirPath = path.join(cwd, defaultD1DirPath)
+
+  const d1Databases = await glob(path.join(d1DirPath, '*.sqlite'), {})
+
+  if (d1Databases.length === 0) {
+    throw new Error(
+      `No Cloudflare D1 databases found in ${defaultD1DirPath}. Did you run \`wrangler d1 create <DATABASE_NAME>\` and \`wrangler dev\`?`,
```
Comment (jkomyno) [inline packages/internals/src/utils/cloudflareD1.ts:24]: I had originally taken the `wrangler dev` bit from https://github.com/prisma/prisma/pull/23613.
However, that doesn't seem to be enough.

The minimum effort required to generate a local D1 database on disk (besides `wrangler d1 create __YOUR_DATABASE_NAME__`) is to run at least a query on such local database. For example, we can ask users to run:

```
npx wrangler d1 execute __YOUR_DATABASE_NAME__ --local --command "PRAGMA table_list"
```

which shows:

```
 ⛅️ wrangler 3.39.0 (update available 3.42.0)
-------------------------------------------------------
🌀 Mapping SQL input into an array of statements
🌀 Executing on local database __YOUR_DATABASE_NAME__ (1cf2f856-fefd-46e7-a888-f192a6edc6aa) from .wrangler/state/v3/d1:
🌀 To execute on your remote database, add a --remote flag to your wrangler command.
┌────────┬────────────────────┬───────┬──────┬────┬────────┐
│ schema │ name               │ type  │ ncol │ wr │ strict │
├────────┼────────────────────┼───────┼──────┼────┼────────┤
│ main   │ _cf_KV             │ table │ 2    │ 1  │ 0      │
├────────┼────────────────────┼───────┼──────┼────┼────────┤
│ main   │ sqlite_schema      │ table │ 5    │ 0  │ 0      │
├────────┼────────────────────┼───────┼──────┼────┼────────┤
│ temp   │ sqlite_temp_schema │ table │ 5    │ 0  │ 0      │
└────────┴────────────────────┴───────┴──────┴────┴────────┘
```

Comment (jkomyno) [reply packages/internals/src/utils/cloudflareD1.ts:24]: Note: we can still merge this as is. As @nikolasburk said, we can assume that the local database already exists when our users try running `migrate diff --from-local-d1` or `--to-local-d1`.

File: packages/migrate/src/commands/DbPull.ts [hunk 3]
```diff
@@ -187,6 +193,30 @@ Set composite types introspection depth to 2 levels
...
+            })
+
+            const result = { firstDatasource: config.datasources[0], schema }
+
+            const hasDriverAdaptersPreviewFeature = (previewFeatures || []).includes('driverAdapters')
+            const validationWarning = `Without the ${bold(
+              'driverAdapters',
+            )} preview feature, the schema introspected via the ${bold('--local-d1')} flag will not work with ${bold(
+              '@prisma/client',
+            )}.`
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbPull.ts:213]: This is weird 🤔

File: packages/migrate/src/commands/MigrateDiff.ts [hunk 4]
```diff
@@ -42,6 +43,11 @@ ${italic('From and To inputs (1 `--from-...` and 1 `--to-...` must be provided):
   --from-migrations        Path to the Prisma Migrate migrations directory
   --to-migrations
 
+  --from-local-d1          Automatically locate the local Cloudflare D1 database, see ${link(
+    'https://pris.ly/d/diff-cloudflare-d1',
```
Comment (Jolg42) [inline packages/migrate/src/commands/MigrateDiff.ts:47]: This does no exists yet?

**Review by Jolg42 — APPROVED** [review]
Nice 🙌🏼

File: packages/migrate/src/__tests__/DbPull/__snapshots__/sqlite.test.ts.snap (hunk 1 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/DbPull/__snapshots__/sqlite.test.ts.snap:174]: Nice!

File: packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap [hunk 5]
```diff
@@ -1,5 +1,87 @@
...
+  --from-schema-datamodel  Path to a Prisma schema file, uses the datamodel for the diff
+  --to-schema-datamodel
+
+  --from-schema-datasource Path to a Prisma schema file, uses the datasource url for the diff
+  --to-schema-datasource
+
+  --from-migrations        Path to the Prisma Migrate migrations directory
+  --to-migrations
+
+  --from-local-d1          Automatically locate the local Cloudflare D1 database, see https://pris.ly/d/diff-cloudflare-d1
```
Comment (jkomyno) [inline packages/migrate/src/__tests__/__snapshots__/MigrateDiff.test.ts.snap:31]:
```suggestion
  --from-local-d1          Automatically locate the local Cloudflare D1 database
```

File: packages/migrate/src/commands/MigrateDiff.ts (hunk 4 above)
Comment (jkomyno) [reply packages/migrate/src/commands/MigrateDiff.ts]: Thanks, removed it in https://github.com/prisma/prisma/pull/23662/commits/a44545f3821c00e1c19f4a8355030217f1874f03.

**Pull request merged**